network/crl.go

```diff
@@ -0,0 +1,380 @@
+		return value, nil
+	}
+
+	log.Debugf("CRL: cache: '%s' not found", key)
+	return nil, errors.New("Key not found")
+}
+
+// Put stores CRL in cache
+func (c *DefaultCRLCache) Put(key string, value []byte) error {
+	c.cache.Store(key, value)
+	log.Debugf("CRL: cache: inserted '%s'", key)
+	return nil
+}
+
+// Remove removes item from cache
+func (c *DefaultCRLCache) Remove(key string) error {
+	c.cache.Delete(key)
+	log.Debugf("CRL: cache: removed '%s'", key)
+	return nil
+}
+
+// ParsedCRLCache is similar to CRLCache, but works on top of it and stores parsed and verified CRLs
+type ParsedCRLCache interface {
+	Get(key string) (*pkix.CertificateList, error)
+	Put(key string, value *pkix.CertificateList) error
+	Remove(key string) error
+}
+
+// PRLRUParsedCRLCache is an implementation of ParsedCRLCache that uses LRU cache inside
+type LRUParsedCRLCache struct {
+	cache lru.Cache
+	mutex sync.RWMutex
+}
+
+// NewLRUParsedCRLCache creates new LRUParsedCRLCache, able to store at most maxEntries values
+func NewLRUParsedCRLCache(maxEntries int) *LRUParsedCRLCache {
+	return &LRUParsedCRLCache{cache: lru.Cache{MaxEntries: maxEntries}}
+}
+
+// Get tries to get CRL from cache, returns error if failed
+func (c *LRUParsedCRLCache) Get(key string) (*pkix.CertificateList, error) {
+	log.Debugf("CRL: LRU cache: loading '%s'", key)
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+
+	value, ok := c.cache.Get(key)
+	if ok {
+		value, ok := value.(*pkix.CertificateList)
+		if !ok {
+			// should never happen since Put() only inserts *pkix.CertificateList
+			return nil, errors.New("Unexpected value of invalid type")
+		}
+		log.Debugf("CRL: LRU cache: '%s' found", key)
+		return value, nil
+	}
+
+	return nil, errors.New("Key not found")
+}
+
+// Put stores CRL in cache
+func (c *LRUParsedCRLCache) Put(key string, value *pkix.CertificateList) error {
+	c.mutex.Lock()
+	c.cache.Add(key, value)
+	c.mutex.Unlock()
+	log.Debugf("CRL: LRU cache: inserted '%s'", key)
+	return nil
+}
+
+// Remove removes item from cache
+func (c *LRUParsedCRLCache) Remove(key string) error {
+	c.mutex.Lock()
+	c.cache.Remove(key)
+	c.mutex.Unlock()
+	log.Debugf("CRL: LRU cache: removed '%s'", key)
+	return nil
+}
+
+// DefaultCRLVerifier is a default implementation of CRLVerifier
+type DefaultCRLVerifier struct {
+	Config      CRLConfig
+	Client      CRLClient
+	Cache       CRLCache
+	ParsedCache ParsedCRLCache
+}
+
+// Tries to find cached CRL, fetches using v.Client if not found, checks the signature of CRL using issuerCert
+func (v DefaultCRLVerifier) getCachedOrFetch(uri string, issuerCert *x509.Certificate) (*pkix.CertificateList, error) {
+	// First, try v.ParsedCache
+	crl, err := v.ParsedCache.Get(v.Config.uri)
+	if crl != nil {
+		if err != nil {
+			// non-empty result + error, should never happen
+			return nil, err
+		}
+
+		return crl, nil
+	}
+
+	// If not found, search in v.Cache, if found then parse, verify and store in v.ParsedCache
```

Comment:
why we use 2 cache instead 1? we fetch data, parse and process it, and put it into "ParsedCache". Why we need to put it into the unparsed cache? which benefits it provides?

Reply:
I agree that this makes things complicated. It's not really obvious which one of those you need:
```go
	Cache       CRLCache
	ParsedCache ParsedCRLCache
```
and what's the difference.

I believe it would be enough to cache only parsed results as that is our happy path. On a cache hit we get `*pkix.CertificateList` right away. On a cache miss we take the slow path: do the network request, parse the results, and cache them only after success.